sycl/source/detail/sycl_mem_obj_t.hpp

```diff
@@ -320,6 +337,8 @@ class __SYCL_EXPORT SYCLMemObjT : public SYCLMemObjI {
   // objects can be released in a deferred manner regardless of whether a host
   // pointer was provided or not.
   bool MIsInternal = false;
+  // The number of graphs which are currently using this memory object.
+  size_t MGraphUseCount = 0;
```

Comment:
I'd prefer using a `shared_ptr` or even reusing an existing one like `MSharedPtrStorage` for this reference counting mechanism if possible. It would be thread-safe might be less error prone.

Reply:
I don't think `shared_ptr` is the best here since we really don't care about the graphs that are using the memory object just the number (or really whether there are any at all). I agree this is not thread-safe though so I've made this an atomic value instead.